# Battle of the Toronto Neighborhoods: The Mexican Restaurant

# Table of Contents

* [Introduction/Business Problem](#intro)
* [Data](#data)
* [Methodology](#methodology)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

# Introduction/Business Problem<a name="intro"></a>

### Introduction
Wikipedia states that (as of 2016) Toronto is Canada’s most populous city and the fourth most in North America. It’s diverse, it’s multicultural, it’s a global hub of commerce and entertainment.. and its profile is always growing. To focalize upon Toronto’s restaurant industry, many have come and gone but there have been a rare and resilient few have come to Toronto and actually thrive in the market. Which is why any hopes of a restaurant business to remain afloat and be successful needs to be analyzed and planned carefully. Even before any notion of looking at a piece of real estate or buying your first piece of cutlery, it is important for stakeholders to be able find all data possible and leverage it to make a sound business decision. 

So, I hope that this project can illustrate how restaurant stakeholders strategize launching and breaking ground for a new restaurant in a metropolitan city like Toronto.

### Business Problem
My client, a wealthy and successfull restauranteur is eager to expand his business operations and brand here in Toronto. They're hoping to create an authentic Mexican eating experience rich in culure and cuisine. Since the city of Toronto is a very competitive market, my client needs insightful data in order to decide if it's good to establish this restuarant in the city and in which neighborhood.

### Interested Audience
Albeit that this is a project about trying to break ground for a Mexican restaurant, I believe that the analyzed data can be relevant for any restauranteur trying to make it in the city of Toronto. As well, with the data science practices and techniques used throughout this project, this can be a useful example and resource for any data science practitioner.

# Data <a name="data"></a>

###  Data Sources
These are the data sources used for this report:

* From Wikipedia, I will pull data of Toronto's neighborhoods. With this data, I will create  data frameworks and geological mappings. Here's the link:
https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

* I will need geological coordinates for each postal code, and that is pulled from this csv file: https://cocl.us/Geospatial_data

* The Foursquare API will be used to obtain location and restaurant  data. The pulled data will provide categorical data that can give further analysis that will determine the selection of a location for the restaurant.

### Data Acquisition and Usage
With the data scraped from Wikipedia, I then put it into dataframes where I cleaned it of 'Not assigned' and non-Toronto data. I then merged it with coordinate data so we can geologically map each neighborhood in python. With these neighborhoods mapped, we can then use the Foursquare API to explore restaurant data for each neighborhood. That way, this will create insights for my client as to which neighborhoods have a food scene and which ones do not. Furthermore, we can also see the frequency of Mexican food in the city, which should also help my client in making deicions to break ground for their restaurant.

# Methodology<a name="methodology"></a>

## Step 1: Wrangling location data from Wikipedia page

### Importing necessary packages for wrangling data and creating dataframes

In [1]:
!pip install beautifulsoup4 
from bs4 import BeautifulSoup                            # package for scraping data
!pip install urllib3 
import urllib3                                           # package for working with urls

import pandas as pd                                      # package to handle dataframes
import numpy as np                                       # package to handle scientific computing
import requests                                          # package to handle http requests
from requests.adapters import HTTPAdapter                # package containing transport adapters that requests  
                                                         # uses to define and maintain connections
from requests.packages.urllib3.util.retry import Retry   # package to support urllib3
from IPython.display import display_html                 # public API for display tools in IPython

### Scraping the Wikipedia page for Toronto geospatial data

In [2]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text

soup = BeautifulSoup(source, 'xml')
table=soup.find('table')

print(soup.title)
tab = str(soup.table)
display_html(tab,raw=True)

<title>List of postal codes of Canada: M - Wikipedia</title>


Postal Code
 
 Borough
 
 Neighborhood
 
 
 M1A
 
 Not assigned
 
 Not assigned
 
 
 M2A
 
 Not assigned
 
 Not assigned
 
 
 M3A
 
 North York
 
 Parkwoods
 
 
 M4A
 
 North York
 
 Victoria Village
 
 
 M5A
 
 Downtown Toronto
 
 Regent Park, Harbourfront
 
 
 M6A
 
 North York
 
 Lawrence Manor, Lawrence Heights
 
 
 M7A
 
 Downtown Toronto
 
 Queen's Park, Ontario Provincial Government
 
 
 M8A
 
 Not assigned
 
 Not assigned
 
 
 M9A
 
 Etobicoke
 
 Islington Avenue, Humber Valley Village
 
 
 M1B
 
 Scarborough
 
 Malvern, Rouge
 
 
 M2B
 
 Not assigned
 
 Not assigned
 
 
 M3B
 
 North York
 
 Don Mills
 
 
 M4B
 
 East York
 
 Parkview Hill, Woodbine Gardens
 
 
 M5B
 
 Downtown Toronto
 
 Garden District, Ryerson
 
 
 M6B
 
 North York
 
 Glencairn
 
 
 M7B
 
 Not assigned
 
 Not assigned
 
 
 M8B
 
 Not assigned
 
 Not assigned
 
 
 M9B
 
 Etobicoke
 
 West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
 
 
 M1C
 
 Scarborough
 
 Rouge Hill, Port Union, Highland Creek
 
 
 M2C
 
 Not assigned
 
 Not assigned
 
 
 M3C
 
 North York
 
 Don Mills
 
 
 M4C
 
 East York
 
 Woodbine Heights
 
 
 M5C
 
 Downtown Toronto
 
 St. James Town
 
 
 M6C
 
 York
 
 Humewood-Cedarvale
 
 
 M7C
 
 Not assigned
 
 Not assigned
 
 
 M8C
 
 Not assigned
 
 Not assigned
 
 
 M9C
 
 Etobicoke
 
 Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
 
 
 M1E
 
 Scarborough
 
 Guildwood, Morningside, West Hill
 
 
 M2E
 
 Not assigned
 
 Not assigned
 
 
 M3E
 
 Not assigned
 
 Not assigned
 
 
 M4E
 
 East Toronto
 
 The Beaches
 
 
 M5E
 
 Downtown Toronto
 
 Berczy Park
 
 
 M6E
 
 York
 
 Caledonia-Fairbanks
 
 
 M7E
 
 Not assigned
 
 Not assigned
 
 
 M8E
 
 Not assigned
 
 Not assigned
 
 
 M9E
 
 Not assigned
 
 Not assigned
 
 
 M1G
 
 Scarborough
 
 Woburn
 
 
 M2G
 
 Not assigned
 
 Not assigned
 
 
 M3G
 
 Not assigned
 
 Not assigned
 
 
 M4G
 
 East York
 
 Leaside
 
 
 M5G
 
 Downtown Toronto
 
 Central Bay Street
 
 
 M6G
 
 Downtown Toronto
 
 Christie
 
 
 M7G
 
 Not assigned
 
 Not assigned
 
 
 M8G
 
 Not assigned
 
 Not assigned
 
 
 M9G
 
 Not assigned
 
 Not assigned
 
 
 M1H
 
 Scarborough
 
 Cedarbrae
 
 
 M2H
 
 North York
 
 Hillcrest Village
 
 
 M3H
 
 North York
 
 Bathurst Manor, Wilson Heights, Downsview North
 
 
 M4H
 
 East York
 
 Thorncliffe Park
 
 
 M5H
 
 Downtown Toronto
 
 Richmond, Adelaide, King
 
 
 M6H
 
 West Toronto
 
 Dufferin, Dovercourt Village
 
 
 M7H
 
 Not assigned
 
 Not assigned
 
 
 M8H
 
 Not assigned
 
 Not assigned
 
 
 M9H
 
 Not assigned
 
 Not assigned
 
 
 M1J
 
 Scarborough
 
 Scarborough Village
 
 
 M2J
 
 North York
 
 Fairview, Henry Farm, Oriole
 
 
 M3J
 
 North York
 
 Northwood Park, York University
 
 
 M4J
 
 East York
 
 East Toronto, Broadview North (Old East York)
 
 
 M5J
 
 Downtown Toronto
 
 Harbourfront East, Union Station, Toronto Islands
 
 
 M6J
 
 West Toronto
 
 Little Portugal, Trinity
 
 
 M7J
 
 Not assigned
 
 Not assigned
 
 
 M8J
 
 Not assigned
 
 Not assigned
 
 
 M9J
 
 Not assigned
 
 Not assigned
 
 
 M1K
 
 Scarborough
 
 Kennedy Park, Ionview, East Birchmount Park
 
 
 M2K
 
 North York
 
 Bayview Village
 
 
 M3K
 
 North York
 
 Downsview
 
 
 M4K
 
 East Toronto
 
 The Danforth West, Riverdale
 
 
 M5K
 
 Downtown Toronto
 
 Toronto Dominion Centre, Design Exchange
 
 
 M6K
 
 West Toronto
 
 Brockton, Parkdale Village, Exhibition Place
 
 
 M7K
 
 Not assigned
 
 Not assigned
 
 
 M8K
 
 Not assigned
 
 Not assigned
 
 
 M9K
 
 Not assigned
 
 Not assigned
 
 
 M1L
 
 Scarborough
 
 Golden Mile, Clairlea, Oakridge
 
 
 M2L
 
 North York
 
 York Mills, Silver Hills
 
 
 M3L
 
 North York
 
 Downsview
 
 
 M4L
 
 East Toronto
 
 India Bazaar, The Beaches West
 
 
 M5L
 
 Downtown Toronto
 
 Commerce Court, Victoria Hotel
 
 
 M6L
 
 North York
 
 North Park, Maple Leaf Park, Upwood Park
 
 
 M7L
 
 Not assigned
 
 Not assigned
 
 
 M8L
 
 Not assigned
 
 Not assigned
 
 
 M9L
 
 North York
 
 Humber Summit
 
 
 

### Create pandas dataframe from the scraped Wikipedia data

In [3]:
dfs = pd.read_html(tab)
df=dfs[0]
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

## Step 2: Cleaning location data & creating and exploring the dataframe

### Dropping the rows where borough is 'Not assigned'

In [4]:
df1 = df[df.Borough != 'Not assigned']
df1.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Combining the neighborhoods with same postal code

In [5]:
df2 = df1.groupby(['Postal Code','Borough'], sort=False).agg(', '.join)
df2.reset_index(inplace=True)
df2.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Replacing the name of neighborhoods which are 'Not assigned' with names of borough

In [6]:
df2['Neighborhood'] = np.where(df2['Neighborhood'] == 'Not assigned',df2['Borough'], df2['Neighborhood'])
df2.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Filtering for only 'Toronto' neighborhoods

In [7]:
df3 = df2[df2['Borough'].str.contains('Toronto',regex=False)]
df3.head()

Postal Code           Borough                                 Neighborhood
2          M5A  Downtown Toronto                    Regent Park, Harbourfront
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
9          M5B  Downtown Toronto                     Garden District, Ryerson
15         M5C  Downtown Toronto                               St. James Town
19         M4E      East Toronto                                  The Beaches

### Shape of 'Toronto' only boroughs data frame

In [8]:
print('The dataframe contains {} boroughs and {} neighborhoods.'.format(
        len(df3['Borough'].unique()),df3.shape[0]))

The dataframe contains 4 boroughs and 39 neighborhoods.


### Looking at neighborhood count by borough

In [9]:
print(df3.groupby('Borough').count()['Neighborhood'])

Borough
Central Toronto      9
Downtown Toronto    19
East Toronto         5
West Toronto         6
Name: Neighborhood, dtype: int64


## Step 3: Adding the longitudes and latitudes to dataset

### Importing csv file containing longitude and latitude data for various Canadian postal codes

In [10]:
long_lat = pd.read_csv('https://cocl.us/Geospatial_data')
long_lat.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Merge the two tables together so that we have the longitude and latitude for all the postal codes from the Part 2 dataset

In [11]:
df4 = pd.merge(df3,long_lat, on = "Postal Code")
df4.tail()

Postal Code           Borough  \
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
34                                     Stn A PO Boxes  43.646435 -79.374846  
35                        St. James Town, Cabbagetown  43.667967 -79.367675  
36             First Canadian Place, Underground city  43.648429 -79.382280  
37                               Church and Wellesley  43.665860 -79.383160  
38  Business reply mail Processing Centre, South C...  43.662744 -79.321558

## Step 4: Visualization and Clustering of Toronto neighborhoods

### Importing packages for geomapping visualizations and clustering

In [12]:
!conda install -c conda-forge geopy --yes                 
from geopy.geocoders import Nominatim                    # module for geocoding
!conda install -c conda-forge folium=0.5.0 --yes         
import folium                                            # module to create leaflet maps

import matplotlib as plt                                 # visualization package
import matplotlib.cm as cm                               # visualization package with colour mapping features
import matplotlib.colors as colors                       # visualization package to display various colors
from sklearn.cluster import KMeans                       # module used to cluster data points

Solving environment: / 
  - anaconda/osx-64::ca-certificates-2020.1.1-0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_4
  - anaconda/osx-64::openssl-1.1.1d-h1de35cc_4, defaults/osx-64::ca-certificates-2020.1.1-0
  - anaconda/osx-64::ca-certificates-2020.1.1-0, defaults/osx-64::openssl-1.1.1d-h1de35cc_4
  - defaults/osx-64::ca-certificates-2020.1.1-0, defaults/osx-64::openssl-1.1.1d-h1de35ccdone

# All requested packages already installed.

Solving environment: / 
  - anaconda/osx-64::ca-certificates-2020.1.1-0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_4
  - anaconda/osx-64::ca-certificates-2020.1.1-0, defaults/osx-64::openssl-1.1.1d-h1de35cc_4
  - anaconda/osx-64::openssl-1.1.1d-h1de35cc_4, defaults/osx-64::ca-certificates-2020.1.1-0
  - defaults/osx-64::ca-certificates-2020.1.1-0, defaults/osx-64::openssl-1.1.1d-h1de35ccdone

# All requested packages already installed.



### Use Geopy to get longitude and latitude of Toronto for Folium mapping

In [13]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="TO")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


### Creating visualization using the Folium package and Part 3's dataset to create mapping points

In [14]:
map_toronto = folium.Map(location=[43.6534817,-79.3839347],zoom_start=10)

for lat,lng,borough,neighborhood in zip(df4['Latitude'],df4['Longitude'],df4['Borough'],df4['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=4,
    popup=label,
    color='red',
    fill=True,
    fill_color='white',
    fill_opacity=.7,
    parse_html=False).add_to(map_toronto)

map_toronto

### To use kmeans clustering, I based my k value upon the number of boroughs there are in Toronto

In [15]:
df4.Borough.unique()

array(['Downtown Toronto', 'East Toronto', 'West Toronto',
       'Central Toronto'], dtype=object)

### Creating the clusters by finding each neighborhood's best fit

In [16]:
k=4

toronto_clustering = df4.drop(['Postal Code','Borough','Neighborhood'],axis=1)
kmeans = KMeans(n_clusters = k,random_state=0).fit(toronto_clustering)
kmeans.labels_

array([3, 3, 3, 3, 0, 3, 3, 1, 3, 1, 3, 1, 0, 3, 1, 0, 3, 0, 2, 2, 2, 2,
       1, 2, 3, 1, 2, 3, 1, 2, 3, 2, 3, 3, 3, 3, 3, 3, 0], dtype=int32)

### Creating a new dataframe that includes the each neighborhood's  cluster value (Borough ID) 

In [17]:
toronto_cluster = df4.copy()
toronto_cluster["Borough ID"] = kmeans.labels_
toronto_cluster.sort_values(by='Borough')

Postal Code           Borough  \
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
26         M4S   Central Toronto   
29         M4T   Central Toronto   
31         M4V   Central Toronto   
18         M4N   Central Toronto   
30         M5T  Downtown Toronto   
37         M4Y  Downtown Toronto   
27         M5S  Downtown Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
0          M5A  Downtown Toronto   
16         M5L  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
13         M5K  Downtown Toronto   
3          M5C  Downtown Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
10         M5J  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
4          M4E      East Toronto   
38         M7Y      East Toronto   
12         M4K      East Toronto   
15         M4L      East Toronto   
17         M4M      East Toronto   
9          M6H      West Toronto   
25         M6R      West Toronto   
11         M6J      West Toronto   
22         M6P      West Toronto   
14         M6K      West Toronto   
28         M6S      West Toronto   

                                         Neighborhood   Latitude  Longitude  \
19                                           Roselawn  43.711695 -79.416936   
20                                   Davisville North  43.712751 -79.390197   
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
23                  North Toronto West, Lawrence Park  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
26                                         Davisville  43.704324 -79.388790   
29                        Moore Park, Summerhill East  43.689574 -79.383160   
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
18                                      Lawrence Park  43.728020 -79.388790   
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
37                               Church and Wellesley  43.665860 -79.383160   
27                     University of Toronto, Harbord  43.662696 -79.400049   
32  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
33                                           Rosedale  43.679563 -79.377529   
34                                     Stn A PO Boxes  43.646435 -79.374846   
35                        St. James Town, Cabbagetown  43.667967 -79.367675   
36             First Canadian Place, Underground city  43.648429 -79.382280   
0                           Regent Park, Harbourfront  43.654260 -79.360636   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
3                                      St. James Town  43.651494 -79.375418   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
4                                         The Beaches  43.676357 -79.293031   
38  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
15                     India Bazaa

### Creating visualization of Toronto's boroughs based on kmeans clusters

In [18]:
# create map
map_clusters = folium.Map(location=[43.6534817,-79.3839347],zoom_start=10)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.Dark2(np.linspace(0, 1, len(ys)))
palette = [colors.rgb2hex(i) for i in colors_array]


# add markers to the map while each cluster/borough is defined by a color
markers_colors = []
for lat, lon, neighbourhood, cluster in zip(toronto_cluster['Latitude'], toronto_cluster['Longitude'], toronto_cluster['Neighborhood'], toronto_cluster['Borough ID']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=palette[cluster-1],
        fill=True,
        fill_color=palette[cluster-1],
        fill_opacity=.7, 
        parse_html=False).add_to(map_clusters)
    
map_clusters

## Step 5: Using Foursquare to Explore Neighborhoods

### Import packages to handle json files

In [19]:
import json                                        # library to handle JSON files
from pandas.io.json import json_normalize          # package that tranforms JSON file into a pandas dataframe

### Defining my Foursquare credentials

In [20]:
CLIENT_ID = 'MDJKWOUGU5UE4VAT1W3GCSDCLHX1TOX3KVVOMKECGCEXTKSU' 
CLIENT_SECRET = 'A33NRTEDKWQUV0SPJN3YKVSKOBDW2OXNLLGWYEHACYJ12HHD' 
VERSION = '20180605' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MDJKWOUGU5UE4VAT1W3GCSDCLHX1TOX3KVVOMKECGCEXTKSU
CLIENT_SECRET:A33NRTEDKWQUV0SPJN3YKVSKOBDW2OXNLLGWYEHACYJ12HHD


### Defining Toronto's geolocation coordinates

In [21]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="TO_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Creating a json file of Foursquare venue data

In [22]:
radius = 1000
LIMIT = 100

def getVenues(names, latitudes, longitudes):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):

# create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION,
            lat, 
            lng, 
            radius, 
            LIMIT)
            
# make the GET request
        venue_results = requests.get(url).json()['response']['groups'][0]['items']
        
# return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood_Latitude', 
                  'Neighborhood_Longitude', 
                  'Venue', 
                  'Venue_Latitude', 
                  'Venue_Longitude', 
                  'Venue_Category']
    
    return(nearby_venues)

### Creating dataframe of Toronto's 'nearby restaurants'

In [25]:
to_venues = getVenues(names=toronto_cluster['Neighborhood'],
                            latitudes=toronto_cluster['Latitude'],
                            longitudes=toronto_cluster['Longitude'])
to_venues

Neighborhood  \
0                             Regent Park, Harbourfront   
1                             Regent Park, Harbourfront   
2                             Regent Park, Harbourfront   
3                             Regent Park, Harbourfront   
4                             Regent Park, Harbourfront   
...                                                 ...   
3191  Business reply mail Processing Centre, South C...   
3192  Business reply mail Processing Centre, South C...   
3193  Business reply mail Processing Centre, South C...   
3194  Business reply mail Processing Centre, South C...   
3195  Business reply mail Processing Centre, South C...   

      Neighborhood_Latitude  Neighborhood_Longitude  \
0                 43.654260              -79.360636   
1                 43.654260              -79.360636   
2                 43.654260              -79.360636   
3                 43.654260              -79.360636   
4                 43.654260              -79.360636   
...                     ...                     ...   
3191              43.662744              -79.321558   
3192              43.662744              -79.321558   
3193              43.662744              -79.321558   
3194              43.662744              -79.321558   
3195              43.662744              -79.321558   

                                 Venue  Venue_Latitude  Venue_Longitude  \
0                     Roselle Desserts       43.653447       -79.362017   
1                        Tandem Coffee       43.653559       -79.361809   
2                      Corktown Common       43.655618       -79.356211   
3     The Distillery Historic District       43.650244       -79.359323   
4               Cooper Koo Family YMCA       43.653249       -79.358008   
...                                ...             ...              ...   
3191                      Leslie Jones       43.662960       -79.331834   
3192                       Tim Hortons       43.662644       -79.309945   
3193                    Breakfast Club       43.662811       -79.310174   
3194                   Carters Landing       43.662414       -79.309898   
3195  Ashbridges Bay Yacht Club (ABYC)       43.658210       -79.311004   

           Venue_Category  
0                  Bakery  
1             Coffee Shop  
2                    Park  
3           Historic Site  
4     Distribution Center  
...                   ...  
3191  American Restaurant  
3192          Coffee Shop  
3193       Breakfast Spot  
3194               Bistro  
3195      Harbor / Marina  

[3196 rows x 7 columns]

In [53]:
tor_venues = to_venues[to_venues['Venue_Category'].str.contains('staurant')]
tor_venues.head()

Neighborhood  Neighborhood_Latitude  Neighborhood_Longitude  \
6   Regent Park, Harbourfront               43.65426              -79.360636   
10  Regent Park, Harbourfront               43.65426              -79.360636   
20  Regent Park, Harbourfront               43.65426              -79.360636   
24  Regent Park, Harbourfront               43.65426              -79.360636   
31  Regent Park, Harbourfront               43.65426              -79.360636   

                         Venue  Venue_Latitude  Venue_Longitude  \
6               Impact Kitchen       43.656369       -79.356980   
10                 Souk Tabule       43.653756       -79.354390   
20  Cluny Bistro & Boulangerie       43.650565       -79.357843   
24   Mangia and Bevi Resto-Bar       43.652250       -79.366355   
31                   Sukhothai       43.658444       -79.365681   

              Venue_Category  
6                 Restaurant  
10  Mediterranean Restaurant  
20         French Restaurant  
24        Italian Restaurant  
31           Thai Restaurant

In [55]:
print('There are {} restaurants in the city of Toronto.'.format(
    tor_venues.Venue_Category.count()))

There are 789 restaurants in the city of Toronto.


In [56]:
print('There are {} uniques restaurant cuisines.'.format(len(tor_venues['Venue_Category'].unique())))

There are 53 uniques restaurant cuisines.


### Creating 2 dataframes where one has only Mexican restaurants while the other has zero Mexican restaurants

In [57]:
MexTO = tor_venues.loc[tor_venues['Venue_Category'] == 'Mexican Restaurant']
MexTO.head()

Neighborhood  Neighborhood_Latitude  \
114  Queen's Park, Ontario Provincial Government              43.662301   
227                     Garden District, Ryerson              43.657162   
415                                  The Beaches              43.676357   
640                           Central Bay Street              43.657952   
657                           Central Bay Street              43.657952   

     Neighborhood_Longitude                   Venue  Venue_Latitude  \
114              -79.389494            Como En Casa       43.665160   
227              -79.378937  Chipotle Mexican Grill       43.656860   
415              -79.293031                    Xola       43.672603   
640              -79.387383  Chipotle Mexican Grill       43.656860   
657              -79.387383            Como En Casa       43.665160   

     Venue_Longitude      Venue_Category  
114       -79.384796  Mexican Restaurant  
227       -79.380910  Mexican Restaurant  
415       -79.288080  Mexican Restaurant  
640       -79.380910  Mexican Restaurant  
657       -79.384796  Mexican Restaurant

In [84]:
noMexTO = tor_venues.loc[tor_venues['Venue_Category'] != 'Mexican Restaurant']
noMexTO.head()

Neighborhood  Neighborhood_Latitude  Neighborhood_Longitude  \
6   Regent Park, Harbourfront               43.65426              -79.360636   
10  Regent Park, Harbourfront               43.65426              -79.360636   
20  Regent Park, Harbourfront               43.65426              -79.360636   
24  Regent Park, Harbourfront               43.65426              -79.360636   
31  Regent Park, Harbourfront               43.65426              -79.360636   

                         Venue  Venue_Latitude  Venue_Longitude  \
6               Impact Kitchen       43.656369       -79.356980   
10                 Souk Tabule       43.653756       -79.354390   
20  Cluny Bistro & Boulangerie       43.650565       -79.357843   
24   Mangia and Bevi Resto-Bar       43.652250       -79.366355   
31                   Sukhothai       43.658444       -79.365681   

              Venue_Category  
6                 Restaurant  
10  Mediterranean Restaurant  
20         French Restaurant  
24        Italian Restaurant  
31           Thai Restaurant

## Step 6: Clustering Foursquare data 

### Create a dataframe of the top 10 venue types of each neighborhood that do not have a Mexican restaurant. We do this by using one hot enoding to turn the restaurant type into numerical data

In [59]:
# one hot encoding
to_onehot = pd.get_dummies(tor_venues[['Venue_Category']], prefix="", prefix_sep="")

# add neighborhood column into dataframe
to_onehot['Neighborhood'] = tor_venues['Neighborhood'] 

# create dataframe with venue common mean by Neighborhood
to_rank = to_onehot.groupby('Neighborhood').mean().reset_index()
print(to_rank.shape)
to_rank

(37, 54)


Neighborhood  American Restaurant  \
0                                         Berczy Park             0.055556   
1        Brockton, Parkdale Village, Exhibition Place             0.038462   
2   Business reply mail Processing Centre, South C...             0.111111   
3                                  Central Bay Street             0.041667   
4                                            Christie             0.029412   
5                                Church and Wellesley             0.043478   
6                      Commerce Court, Victoria Hotel             0.043478   
7                                          Davisville             0.000000   
8                                    Davisville North             0.000000   
9                        Dufferin, Dovercourt Village             0.000000   
10             First Canadian Place, Underground city             0.047619   
11    Forest Hill North & West, Forest Hill Road Park             0.000000   
12                           Garden District, Ryerson             0.035714   
13  Harbourfront East, Union Station, Toronto Islands             0.000000   
14                      High Park, The Junction South             0.000000   
15                     India Bazaar, The Beaches West             0.000000   
16          Kensington Market, Chinatown, Grange Park             0.000000   
17                           Little Portugal, Trinity             0.030303   
18                        Moore Park, Summerhill East             0.058824   
19                  North Toronto West, Lawrence Park             0.000000   
20                             Parkdale, Roncesvalles             0.090909   
21        Queen's Park, Ontario Provincial Government             0.000000   
22                          Regent Park, Harbourfront             0.000000   
23                           Richmond, Adelaide, King             0.052632   
24                                           Rosedale             0.000000   
25                                           Roselawn             0.000000   
26                                 Runnymede, Swansea             0.062500   
27                                     St. James Town             0.040000   
28                        St. James Town, Cabbagetown             0.100000   
29                                     Stn A PO Boxes             0.090909   
30                                    Studio District             0.142857   
31  Summerhill West, Rathnelly, South Hill, Forest...             0.041667   
32                The Annex, North Midtown, Yorkville             0.032258   
33                                        The Beaches             0.000000   
34                       The Danforth West, Riverdale             0.031250   
35           Toronto Dominion Centre, Design Exchange             0.043478   
36                     University of Toronto, Harbord             0.000000   

    Asian Restaurant  Belgian Restaurant  Brazilian Restaurant  \
0           0.000000            0.000000              0.000000   
1           0.000000            0.000000              0.000000   
2           0.000000            0.000000              0.000000   
3           0.000000            0.000000              0.000000   
4           0.000000            0.000000              0.000000   
5           0.000000            0.000000              0.000000   
6           0.000000            0.000000              0.000000   
7           0.000000            0.000000              0.000000   
8           0.000000            0.000000              0.000000   
9           0.000000            0.000000              0.133333   
10          0.000000            0.000000              0.047619   
11          0.000000            0.000000              0.000000   
12          0.000000            0.000000              0.000000   
13          0.000000            0.000000              0.000000   
14          0.000000            0.000000              0.000000   
15          0.043478

### Creating printout of each neighborhood ranking it's top 5 most common restaurant types

In [60]:
num_top_venues = 5

for hood in to_rank['Neighborhood']:
    print("----"+hood+"----")
    temp = to_rank[to_rank['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                           venue  freq
0            Japanese Restaurant  0.28
1                     Restaurant  0.22
2              French Restaurant  0.06
3  Vegetarian / Vegan Restaurant  0.06
4                Thai Restaurant  0.06


----Brockton, Parkdale Village, Exhibition Place----
                           venue  freq
0                     Restaurant  0.23
1             Tibetan Restaurant  0.12
2  Vegetarian / Vegan Restaurant  0.08
3             Italian Restaurant  0.08
4              Indian Restaurant  0.08


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0  Fast Food Restaurant  0.22
1    Italian Restaurant  0.22
2      Sushi Restaurant  0.22
3   American Restaurant  0.11
4       Thai Restaurant  0.11


----Central Bay Street----
                           venue  freq
0               Sushi Restaurant  0.17
1            Japanese Restaurant  0.12
2               Ramen Restaurant  0.

### Creating a dataframe of each neighborhood's top 10 most common restaurant types

In [61]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [62]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Neighborhood'] = to_rank['Neighborhood']

for ind in np.arange(to_rank.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(to_rank.iloc[ind, :], num_top_venues)

venues_sorted

Neighborhood  \
0                                         Berczy Park   
1        Brockton, Parkdale Village, Exhibition Place   
2   Business reply mail Processing Centre, South C...   
3                                  Central Bay Street   
4                                            Christie   
5                                Church and Wellesley   
6                      Commerce Court, Victoria Hotel   
7                                          Davisville   
8                                    Davisville North   
9                        Dufferin, Dovercourt Village   
10             First Canadian Place, Underground city   
11    Forest Hill North & West, Forest Hill Road Park   
12                           Garden District, Ryerson   
13  Harbourfront East, Union Station, Toronto Islands   
14                      High Park, The Junction South   
15                     India Bazaar, The Beaches West   
16          Kensington Market, Chinatown, Grange Park   
17                           Little Portugal, Trinity   
18                        Moore Park, Summerhill East   
19                  North Toronto West, Lawrence Park   
20                             Parkdale, Roncesvalles   
21        Queen's Park, Ontario Provincial Government   
22                          Regent Park, Harbourfront   
23                           Richmond, Adelaide, King   
24                                           Rosedale   
25                                           Roselawn   
26                                 Runnymede, Swansea   
27                                     St. James Town   
28                        St. James Town, Cabbagetown   
29                                     Stn A PO Boxes   
30                                    Studio District   
31  Summerhill West, Rathnelly, South Hill, Forest...   
32                The Annex, North Midtown, Yorkville   
33                                        The Beaches   
34                       The Danforth West, Riverdale   
35           Toronto Dominion Centre, Design Exchange   
36                     University of Toronto, Harbord   

            1st Most Common Venue          2nd Most Common Venue  \
0             Japanese Restaurant                     Restaurant   
1                      Restaurant             Tibetan Restaurant   
2              Italian Restaurant           Fast Food Restaurant   
3                Sushi Restaurant            Japanese Restaurant   
4               Korean Restaurant             Mexican Restaurant   
5             Japanese Restaurant               Sushi Restaurant   
6                      Restaurant            Japanese Restaurant   
7              Italian Restaurant               Sushi Restaurant   
8              Italian Restaurant           Fast Food Restaurant   
9                Sushi Restaurant             Italian Restaurant   
10                     Restaurant            Japanese Restaurant   
11               Sushi Restaurant            Japanese Restaurant   
12            Japanese Restaurant             Italian Restaurant   
13            Japanese Restaurant             Italian Restaurant   
14             Italian Restaurant                Thai Restaurant   
15              Indian Restaurant           Fast Food Restaurant   
16  Vegetarian / Vegan Restaurant             Mexican Restaurant   
17                     Restaurant  Vegetarian / Vegan Restaurant   
18             Italian Restaurant                Thai Restaurant   
19             Italian Restaurant             Mexican Restaurant   
20                Thai Restaurant               Sushi Restaurant   
21            Japanese Restaurant             Italian Restaurant   
22             Italian Restaurant                     Restaurant   
23            Japanese Restaurant                     Restaurant   
24            Japanese Restaurant              Indian Restaurant   
25               Sushi Restaurant             Italian Restaurant   
26             Italian Restaurant     

## Step 7: Clustering Toronto neighborhood venues

### Using the previous k value of 4, I created clusters based upon the Foursquare with neighborhoods that do not have a Mexican restaurant

In [63]:
k_clusters = 4

to_fsquare_clustering = to_rank.drop('Neighborhood', 1)

# run k-means clustering
kmeans1 = KMeans(init = "k-means++", n_clusters=k_clusters, random_state=None).fit(to_fsquare_clustering)

# check cluster labels generated for each row in the dataframe
kmeans1.labels_

array([1, 1, 2, 0, 0, 0, 1, 2, 2, 2, 1, 2, 0, 1, 2, 3, 0, 1, 2, 2, 2, 0,
       2, 1, 3, 2, 2, 1, 0, 1, 2, 2, 0, 0, 0, 1, 0], dtype=int32)

### Merging the clustered foursquare array with the sorted venue data

In [64]:
# add clustering labels
venues_sorted.insert(0, 'Cluster_ID', kmeans1.labels_ )

toronto_rank = noMexTO

# merge toronto_ with neighbourhoods to add latitude/longitude for each neighborhood
to_merged = df4.join(venues_sorted.set_index('Neighborhood'), on='Neighborhood')

to_merged

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
11                           Little Portugal, Trinity  43.647927 -79.419750   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                    Studio District  43.659526 -79.340923   
18                                      Lawrence Park  43.728020 -79.388790   
19                                           Roselawn  43.711695 -79.416936   
20                                   Davisville North  43.712751 -79.390197   
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
22                      High Park, The Junction South  43.661608 -79.464763   
23                  North Toronto West, Lawrence Park  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                             Parkdale, Roncesvalles  43.648960 -79.456325   
26                                         Davisville  43.704324 -79.388790   
27                     University of Toronto, Harbord  43.662696 -79.400049   
28                                 Runnymede, Swansea  43.651571 -79.484450   
29                        Moore Park, Summerhill East  43.689574 -79.383160   
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
31  Summerhill West, Rathnelly, So

### Dropping all 'NaN' data 

In [65]:
to_merged.dropna(axis=0, how='any',inplace=True)
to_merged.reset_index(inplace=True, drop=True)
to_merged

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M5N   Central Toronto   
19         M4P   Central Toronto   
20         M5P   Central Toronto   
21         M6P      West Toronto   
22         M4R   Central Toronto   
23         M5R   Central Toronto   
24         M6R      West Toronto   
25         M4S   Central Toronto   
26         M5S  Downtown Toronto   
27         M6S      West Toronto   
28         M4T   Central Toronto   
29         M5T  Downtown Toronto   
30         M4V   Central Toronto   
31         M4W  Downtown Toronto   
32         M5W  Downtown Toronto   
33         M4X  Downtown Toronto   
34         M5X  Downtown Toronto   
35         M4Y  Downtown Toronto   
36         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
11                           Little Portugal, Trinity  43.647927 -79.419750   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                    Studio District  43.659526 -79.340923   
18                                           Roselawn  43.711695 -79.416936   
19                                   Davisville North  43.712751 -79.390197   
20    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
21                      High Park, The Junction South  43.661608 -79.464763   
22                  North Toronto West, Lawrence Park  43.715383 -79.405678   
23                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
24                             Parkdale, Roncesvalles  43.648960 -79.456325   
25                                         Davisville  43.704324 -79.388790   
26                     University of Toronto, Harbord  43.662696 -79.400049   
27                                 Runnymede, Swansea  43.651571 -79.484450   
28                        Moore Park, Summerhill East  43.689574 -79.383160   
29          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
30  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
31                                           Rosedale  43.679563 -79.377529   
32                         

### Visualizing restaurant clusters but 'black' out areas that already have a Mexican restaurant

In [67]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_clusters)
ys = [i + x + (i*x)**2 for i in range(k_clusters)]
colors_array = cm.Dark2(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(to_merged['Latitude'], to_merged['Longitude'], to_merged['Neighborhood'], to_merged['Cluster_ID']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=6,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)


for lat, lon, poi in zip(MexTO['Venue_Latitude'], MexTO['Venue_Longitude'], MexTO['Neighborhood']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='black',
        fill=True,
        fill_color='black',
        fill_opacity=1).add_to(map_clusters)    
    

map_clusters

## Step 8: Exploring the clusters

In [86]:
GrayPts = to_merged.loc[to_merged['Cluster_ID'] == 0, to_merged.columns[[1] + list(range(2, to_merged.shape[1]))]]
GrayPts

Borough                                 Neighborhood   Latitude  \
1   Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2   Downtown Toronto                     Garden District, Ryerson  43.657162   
4       East Toronto                                  The Beaches  43.676357   
6   Downtown Toronto                           Central Bay Street  43.657952   
7   Downtown Toronto                                     Christie  43.669542   
12      East Toronto                 The Danforth West, Riverdale  43.679557   
23   Central Toronto          The Annex, North Midtown, Yorkville  43.672710   
26  Downtown Toronto               University of Toronto, Harbord  43.662696   
29  Downtown Toronto    Kensington Market, Chinatown, Grange Park  43.653206   
33  Downtown Toronto                  St. James Town, Cabbagetown  43.667967   
35  Downtown Toronto                         Church and Wellesley  43.665860   

    Longitude  Cluster_ID          1st Most Common Venue  \
1  -79.389494         0.0            Japanese Restaurant   
2  -79.378937         0.0            Japanese Restaurant   
4  -79.293031         0.0            Japanese Restaurant   
6  -79.387383         0.0               Sushi Restaurant   
7  -79.422564         0.0              Korean Restaurant   
12 -79.352188         0.0               Greek Restaurant   
23 -79.405678         0.0             Italian Restaurant   
26 -79.400049         0.0  Vegetarian / Vegan Restaurant   
29 -79.400049         0.0  Vegetarian / Vegan Restaurant   
33 -79.367675         0.0            Japanese Restaurant   
35 -79.383160         0.0            Japanese Restaurant   

            2nd Most Common Venue  3rd Most Common Venue  \
1              Italian Restaurant       Ramen Restaurant   
2              Italian Restaurant       Ramen Restaurant   
4            Caribbean Restaurant             Restaurant   
6             Japanese Restaurant       Ramen Restaurant   
7              Mexican Restaurant   Ethiopian Restaurant   
12           Fast Food Restaurant     Italian Restaurant   
23  Vegetarian / Vegan Restaurant             Restaurant   
26             Mexican Restaurant             Restaurant   
29             Mexican Restaurant  Vietnamese Restaurant   
33           Taiwanese Restaurant   Caribbean Restaurant   
35               Sushi Restaurant             Restaurant   

            4th Most Common Venue          5th Most Common Venue  \
1                Sushi Restaurant                Thai Restaurant   
2                      Restaurant      Middle Eastern Restaurant   
4   Vegetarian / Vegan Restaurant               Greek Restaurant   
6              Italian Restaurant  Vegetarian / Vegan Restaurant   
7               Indian Restaurant  Vegetarian / Vegan Restaurant   
12               Ramen Restaurant                     Restaurant   
23                Thai Restaurant              Indian Restaurant   
26            Japanese Restaurant        Comfort Food Restaurant   
29           Caribbean Restaurant              French Restaurant   
33              Indian Restaurant             Italian Restaurant   
35           Caribbean Restaurant                Thai Restaurant   

        6th Most Common Venue     7th Most Common Venue  \
1   Middle Eastern Restaurant                Restaurant   
2          Falafel Restaurant      Fast Food Restaurant   
4           French Restaurant  Mediterranean Restaurant   
6          Mexican Restaurant                Restaurant   
7       Vietnamese Restaurant          Ramen Restaurant   
12        Japanese Restaurant          Asian Restaurant   
23         Mexican Restaurant       Japanese Restaurant   
26         Italian Restaurant        Persian Restaurant   
29                 Restaurant        Belgian Restaurant   
33                 Restaurant          Sushi Restaurant   
35         Italian Restaurant          Ramen Restaurant   

         8th Most Common Venue 9th Most Common Venue  \
1             Theme Restaura

In [73]:
GreenPts = to_merged.loc[to_merged['Cluster_ID'] == 1, to_merged.columns[[1] + list(range(2, to_merged.shape[1]))]]
GreenPts

Borough                                       Neighborhood  \
3   Downtown Toronto                                     St. James Town   
5   Downtown Toronto                                        Berczy Park   
8   Downtown Toronto                           Richmond, Adelaide, King   
10  Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
11      West Toronto                           Little Portugal, Trinity   
13  Downtown Toronto           Toronto Dominion Centre, Design Exchange   
14      West Toronto       Brockton, Parkdale Village, Exhibition Place   
16  Downtown Toronto                     Commerce Court, Victoria Hotel   
32  Downtown Toronto                                     Stn A PO Boxes   
34  Downtown Toronto             First Canadian Place, Underground city   

     Latitude  Longitude  Cluster_ID 1st Most Common Venue  \
3   43.651494 -79.375418         1.0            Restaurant   
5   43.644771 -79.373306         1.0   Japanese Restaurant   
8   43.650571 -79.384568         1.0   Japanese Restaurant   
10  43.640816 -79.381752         1.0   Japanese Restaurant   
11  43.647927 -79.419750         1.0            Restaurant   
13  43.647177 -79.381576         1.0   Japanese Restaurant   
14  43.636847 -79.428191         1.0            Restaurant   
16  43.648198 -79.379817         1.0            Restaurant   
32  43.646435 -79.374846         1.0   Japanese Restaurant   
34  43.648429 -79.382280         1.0            Restaurant   

            2nd Most Common Venue          3rd Most Common Venue  \
3             Japanese Restaurant             Italian Restaurant   
5                      Restaurant             Seafood Restaurant   
8                      Restaurant             Italian Restaurant   
10             Italian Restaurant  Vegetarian / Vegan Restaurant   
11  Vegetarian / Vegan Restaurant               Asian Restaurant   
13                     Restaurant             Italian Restaurant   
14             Tibetan Restaurant              Indian Restaurant   
16            Japanese Restaurant             Seafood Restaurant   
32                     Restaurant             Seafood Restaurant   
34            Japanese Restaurant             Seafood Restaurant   

            4th Most Common Venue          5th Most Common Venue  \
3              Seafood Restaurant        Comfort Food Restaurant   
5         Comfort Food Restaurant              French Restaurant   
8                Sushi Restaurant             Seafood Restaurant   
10                     Restaurant             Chinese Restaurant   
11             Italian Restaurant          Vietnamese Restaurant   
13             Seafood Restaurant                Thai Restaurant   
14             Italian Restaurant  Vegetarian / Vegan Restaurant   
16  Vegetarian / Vegan Restaurant                Thai Restaurant   
32            American Restaurant        New American Restaurant   
34             Italian Restaurant                Thai Restaurant   

            6th Most Common Venue    7th Most Common Venue  \
3            Fast Food Restaurant        French Restaurant   
5                Greek Restaurant       Italian Restaurant   
8            Brazilian Restaurant     Fast Food Restaurant   
10               Greek Restaurant        French Restaurant   
11             Seafood Restaurant      Japanese Restaurant   
13  Vegetarian / Vegan Restaurant     Brazilian Restaurant   
14            Japanese Restaurant  New American Restaurant   
16              French Restaurant       Italian Restaurant   
32               Greek Restaurant          Thai Restaurant   
34  Vegetarian / Vegan Restaurant     Brazilian Restaurant   

            8th Most Common Venue          9th Most Common Venue  \
3               German Restaurant  Vegetarian / Vegan Restaurant   
5   Vegetarian / Vegan Restaurant      Middle Eastern Restaurant   
8   Vegetarian / Vegan Restaurant       Mediterranean Restaurant   
10       Mediterranean Restaurant             Seafo

In [74]:
PurplePts = to_merged.loc[to_merged['Cluster_ID'] == 2, to_merged.columns[[1] + list(range(2, to_merged.shape[1]))]]
PurplePts

Borough                                       Neighborhood  \
0   Downtown Toronto                          Regent Park, Harbourfront   
9       West Toronto                       Dufferin, Dovercourt Village   
17      East Toronto                                    Studio District   
18   Central Toronto                                           Roselawn   
19   Central Toronto                                   Davisville North   
20   Central Toronto    Forest Hill North & West, Forest Hill Road Park   
21      West Toronto                      High Park, The Junction South   
22   Central Toronto                  North Toronto West, Lawrence Park   
24      West Toronto                             Parkdale, Roncesvalles   
25   Central Toronto                                         Davisville   
27      West Toronto                                 Runnymede, Swansea   
28   Central Toronto                        Moore Park, Summerhill East   
30   Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
36      East Toronto  Business reply mail Processing Centre, South C...   

     Latitude  Longitude  Cluster_ID  1st Most Common Venue  \
0   43.654260 -79.360636         2.0     Italian Restaurant   
9   43.669005 -79.442259         2.0       Sushi Restaurant   
17  43.659526 -79.340923         2.0  Vietnamese Restaurant   
18  43.711695 -79.416936         2.0       Sushi Restaurant   
19  43.712751 -79.390197         2.0     Italian Restaurant   
20  43.696948 -79.411307         2.0       Sushi Restaurant   
21  43.661608 -79.464763         2.0     Italian Restaurant   
22  43.715383 -79.405678         2.0     Italian Restaurant   
24  43.648960 -79.456325         2.0        Thai Restaurant   
25  43.704324 -79.388790         2.0     Italian Restaurant   
27  43.651571 -79.484450         2.0     Italian Restaurant   
28  43.689574 -79.383160         2.0     Italian Restaurant   
30  43.686412 -79.400049         2.0       Sushi Restaurant   
36  43.662744 -79.321558         2.0     Italian Restaurant   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0             Restaurant     Indian Restaurant             Sushi Restaurant   
9     Italian Restaurant  Brazilian Restaurant        Portuguese Restaurant   
17   American Restaurant    Italian Restaurant            French Restaurant   
18    Italian Restaurant   Japanese Restaurant             Asian Restaurant   
19  Fast Food Restaurant            Restaurant             Sushi Restaurant   
20   Japanese Restaurant    Italian Restaurant           Persian Restaurant   
21       Thai Restaurant      Sushi Restaurant           Mexican Restaurant   
22    Mexican Restaurant            Restaurant           Chinese Restaurant   
24      Sushi Restaurant            Restaurant  Eastern European Restaurant   
25      Sushi Restaurant  Fast Food Restaurant            Indian Restaurant   
27      Sushi Restaurant    Falafel Restaurant                   Restaurant   
28       Thai Restaurant      Sushi Restaurant                   Restaurant   
30       Thai Restaurant    Italian Restaurant                   Restaurant   
36  Fast Food Restaurant      Sushi Restaurant              Thai Restaurant   

            5th Most Common Venue          6th Most Common Venue  \
0                 Thai Restaurant      Middle Eastern Restaurant   
9           Vietnamese Restaurant             Mexican Restaurant   
17                Thai Restaurant      Latin American Restaurant   
18           Ethiopian Restaurant              Indian Restaurant   
19             Mexican Restaurant          Vietnamese Restaurant   
20  Vegetarian / Vegan Restaurant      Middle Eastern Restaurant   
21                     Restaurant             Seafood Restaurant   
22           Fast Food Restaurant               Sushi Restaurant   
24            American Restaurant             Chinese Restaurant   
25                     Restaurant      Middle Eastern Restaurant   
27     

In [75]:
YellowPts = to_merged.loc[to_merged['Cluster_ID'] == 3, to_merged.columns[[1] + list(range(2, to_merged.shape[1]))]]
YellowPts

Borough                    Neighborhood   Latitude  Longitude  \
15      East Toronto  India Bazaar, The Beaches West  43.668999 -79.315572   
31  Downtown Toronto                        Rosedale  43.679563 -79.377529   

    Cluster_ID 1st Most Common Venue 2nd Most Common Venue  \
15         3.0     Indian Restaurant  Fast Food Restaurant   
31         3.0   Japanese Restaurant     Indian Restaurant   

   3rd Most Common Venue 4th Most Common Venue          5th Most Common Venue  \
15            Restaurant    Italian Restaurant  Vegetarian / Vegan Restaurant   
31   Filipino Restaurant    Italian Restaurant      Indian Chinese Restaurant   

   6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
15      Halal Restaurant     French Restaurant  Middle Eastern Restaurant   
31   Hawaiian Restaurant      Halal Restaurant           Greek Restaurant   

   9th Most Common Venue     10th Most Common Venue  
15  Pakistani Restaurant  Indian Chinese Restaurant  
31     German Restaurant          French Restaurant

# Results and Discussion <a name="results"></a>

Our analysis dictates that even though Toronto has an abundant amount of restaurants (789, according to Foursquare), there are pockets of low density restaurant data for a lot of the areas. Albeit a new restaurant can vitalize a neighborhood and create a new food scene, this is still a gamble that my client would like to avoid. Looking at the above map, we can see established neighborhoods; but if you notice the black points, those are Mexican restaurants. So, we would also want to avoid the surrounding areas as well.  

Furthermore, if you look at the breakdown of each cluster and their common restaurants for each, you can see that Mexican restaurants aren’t that popular in the city; compare to the most common in asian cuisines (Japanese/Sushi, Korean, Vietnamese, Thai), Middle-Eastern cuisines (Indian, Halal, Pakistani) and Italian looking like the most common. So, adding a Mexican restaurant might give some variety to dining in any of these neighborhoods. 

However, with the 2 criterias mentioned prior to an established food scene and a lack of Mexican restaurants in a cluster, I can recommend that the green cluster is the most optimal to break ground for a new Mexican restaurant. 


# Conclusion <a name="conclusion"></a>

The purpose of this project was to identify areas in Toronto that would facilitate a new Mexican restaurant for my client. By looking at clean datasets and cross referencing them with Foursquare data, albeit a limited collection of data points, I was able to create some insights that my client should consider moving forward. And with the recommendation of the green cluster, it’s not a bad choice in any way. South downtown Toronto is an area that filled with various food options, shopping districts and tourist attractions all placed along the subway line. 

If only they had a Mexican restaurant….
